src/javascript/gameInstance.js

```diff
@@ -194,63 +353,62 @@ export default class GameInstance {
       const elapsedTime = Date.now() - startTime;
       this.playTime = Number(elapsedTime / 1000);
+      this.paused = false;
       if (this.playTime > Number(this.vm.noteSpeedInSec)) {
         if (!this.vm.started || !this.paused) this.resumeGame();
         // this.vm.visualizerInstance.initAllVisualizersIfRequried();
         clearInterval(intervalPrePlay);
         clearInterval(this.intervalPlay);
         this.vm.started = true;
-        this.intervalPlay = setInterval(async () => {
-          const cTime = await this.getCurrentTime();
-          this.playTime = cTime + Number(this.vm.noteSpeedInSec);
-          const lastIdx = this.timeArr.length - 1;
-          // advance time arr idx if it's behind current play time
-          if (this.vm.inEditor && !this.vm.playMode && !this.paused) {
-            const lastIdx = this.timeArrIdx;
-            let idx = this.timeArr.findIndex((e) => e.t >= cTime);
-            if (idx === -1) {
-              const endIdx = this.timeArr.length - 1;
-              idx =
-                this.timeArr[endIdx] && this.timeArr[endIdx].t < cTime
-                  ? endIdx + 1
-                  : 0;
-            }
-            this.timeArrIdx = idx;
-            if (
-              this.vm.showExistingNote &&
-              lastIdx !== idx &&
-              this.timeArr[idx]
-            )
-              this.registerKeyDown(this.timeArr[idx].k, "grey");
-          }
-          // check game end
-          if (
-            !this.vm.inEditor &&
-            this.vm.playMode &&
-            this.timeArrIdx >= lastIdx &&
-            this.timeArr[lastIdx] &&
-            this.playTime > Number(this.timeArr[lastIdx].t) + 5
-          ) {
-            this.vm.gameEnded();
-          }
-        }, 100);
+        this.intervalPlay = setInterval(this.gameTimingLoop.bind(this), 100);
       }
     }, 100);
   }
 
+  async gameTimingLoop() {
+    const cTime = await this.getCurrentTime();
+    this.playTime = cTime + Number(this.vm.noteSpeedInSec);
+    this.currentTime = cTime;
+
+    // check game end
+    const lastIdx = this.timeArr.length - 1;
+    if (
+      !this.vm.inEditor &&
+      this.vm.playMode &&
+      this.timeArrIdx >= lastIdx &&
+      this.timeArr[lastIdx] &&
+      this.playTime > Number(this.timeArr[lastIdx].t) + 5 // FIXME: Number Casting
+    ) {
+      this.vm.gameEnded();
+    }
+  }
+
   getCurrentTime() {
     // it seems that 'getPlayerTime' is async, thus all places calling this func need to await res [help wanted]
     return this.vm.srcMode === "youtube"
       ? this.ytPlayer.getPlayerTime()
       : this.audio.getCurrentTime();
   }
 
+  getNoteTiming() {
+    return this.vm.playMode ? this.playTime : this.currentTime; // in editor mode, time without note drop delay is used.
+  }
+
+  isWithinTime(time) {
+    // check if a note with start time is currently on screen
+    const sec = Number(this.vm.noteSpeedInSec);
+    const current = Number(this.getNoteTiming());
```

Comment:
It is for now, there are data currently stored in the backend which numbers are casted as strings. I'll try to run a program to filter those data from the db to make sure they are in Number format, and then these can be safely removed. Otherwise can't think of a good solution for now.

Reply:
Ah that is a different story because the backend serializes the data as strings. Then yeah you'd probably want to make sure they are numbers when u retrieve them from db (i.e. whenever u invoke the db API).